WW-Digital/reactive-kinesis
# Fix #25 by providing an Akka Stream Source graph stage for Kinesis.

Initial PR #36 has been closed in favour of automated integration testing setup.

This graph stage provides an akka source graph based on the actor model provided by the reactive kinesis consumer.

If a consumer is configured, an akka stream Source can be obtained via a simple Kinesis.source("consumer-name").
Every message that flows through the stream needs to be committed explicitly.

An integration test is added that relies on localstack running on the same node (I did not add the localstack startup logic as part of this PR).

Steps to run the test:

Set this env var, since the localstack does not support CBOR: export AWS_CBOR_DISABLE=true
Start localstack (I use the docker-compose with: docker-compose -f localstack.yml up) in default configuration (Kinesis on port 4568, Dynamo on port 4569)
sbt it:test
It was not as easy to come up with a good integration test, since a Kinesis Source usually never finishes. To have a finite test, I use Flow.take which ends the stream after a defined amount of entries. The interplay of different readers/shard consumers could be different between runs, but the test expectation should be met.

@markglh tests and integration tests have passed several successive test runs on my machine.
- There seems to be a race condition in KCL when the lease table is created/accessed, which made the tests brittle.  The test setup creates the lease table now proactively to remedy this shortcoming.
- KinesisProducerIntegrationSpec was broken and is ignored to minimise the change
- The test kinesis configuration settings have been relaxed. The original settings are good for local development but could lead to problems on not so well equipped or loaded machines (like CI server)

Let me know what you think.
